### ISSUE TITLE
tensor multinomial dependent on shape for some reason

### ISSUE THREAD

**Issue by shriphani** [pytorch/pytorch#12309]
## 🐛 Bug

I drew some multinomial samples from the following distribution tensors:

```
foo1 = torch.from_numpy(np.array([0.25, 0.25, 0.25, 0.25]))
foo2 = torch.from_numpy(np.array([[0.25, 0.25, 0.25, 0.25]]))
foo3 = torch.from_numpy(np.array([[[0.25, 0.25, 0.25, 0.25]]]))
```

Here's a histogram of the entries it produces: (first element is the counts, second is the bins)

`foo1`
```
(array([26, 28, 46]), array([0, 1, 2, 3]))
```

this looks alright:

`foo2`
```
(array([25, 28, 47]), array([0, 1, 2, 3]))
```

also looks alright

`foo3`
```
(array([100,   0,   0]), array([0, 1, 2, 3]))
```

basically it just picked the first element - I don't see why `.multinomial` needs to depend on this dimension of the probabilities tensor.

## Environment
```
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04 LTS
GCC version: (Ubuntu 7.3.0-16ubuntu3) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla K40c
GPU 1: GeForce GTX TITAN X
GPU 2: GeForce GTX TITAN X

Nvidia driver version: 396.24
cuDNN version: Probably one of the following:
/usr/local/cuda-9.0/lib64/libcudnn.so
/usr/local/cuda-9.0/lib64/libcudnn.so.7
/usr/local/cuda-9.0/lib64/libcudnn.so.7.1.4
/usr/local/cuda-9.0/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.13.3)
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] pytorch                   0.4.1           py36_py35_py27__9.0.176_7.1.2_2    pytorch
[conda] torch                     0.4.1                     <pip>
[conda] torchfile                 0.1.0                     <pip>
[conda] torchvision               0.2.1                    py36_1    pytorch
```

**Comment by alicanb** [comment]
Even though [docs](https://pytorch.org/docs/stable/torch.html#torch.multinomial) do not cover the case when `foo.dim()`>2, I think this is a bug (also in master). Fortunately, `torch.distributions.Categorical(foo3).sample((100,))` works on master (not sure about 0.4.1), so you can use that for now.

**Comment by zou3519** [comment]
Could you provide us the line you used to reproduce this behavior?

**Comment by alicanb** [comment]
For me, it was 
```
import torch
import numpy as np
foo3 = torch.from_numpy(np.array([[[0.25, 0.25, 0.25, 0.25]]]))
torch.multinomial(foo3, 10, True)
# tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
```

**Comment by shriphani** [comment]
mine was similar to @alicanb 's snippet.

**Comment by weiyangfb** [comment]
Is this needed to be fixed? will it be sufficient if it errors out when `input` has `dim > 2`? @zou3519  @SsnL

**Comment by SsnL** [comment]
@weiyangfb That SGTM. An alternative and better fix is to support >2D, and sample for each vector in last dim.

**Comment by weiyangfb** [comment]
@SsnL Yeah, that would be very nice. But probably will be better after `multinomial` get migrated to ATen.

**Comment by shriphani** [comment]
Here's an example where I run into this sort of stuff:

```
out, hidden = lstm(input, hidden)
logits = w(out) # is nn.linear
choice = logits.softmax(dim=-1).multinomial() # goes pfft here because of dimension
````

I think the right fix is broadcasting along the dim argument (which doesn't exist atm) no ?

**Comment by gchanan** [comment]
@ifedan to investigate.
